EIPS/eip-5843.md

```diff
@@ -0,0 +1,448 @@
+
+To improve the performance of modular multiplication, values can be expressed in Montgomery form.  This allows for the use of Montgomery modular multiplication which replaces the division by the modulus with a bit-shift and single subtraction in the reduction step.
+
+Operating in Montgomery form requires precomputation of a modulus-specific constant as well as the cost of converting values in normal form to/from Montgomery form.  However, these costs are greatly offset by the savings gained when performing many modular multiplications.
+
+(TODO: awkward transition from previous sentence)
+
+This EIP proposes a new model for performing modular arithmetic in the EVM with odd, fixed moduli.  It can provide cost reductions for existing Ethereum applications, greatly increased utility to the EVM while introducing a mostly self-contained spec which doesn't introduce deep changes to the EVM (TODO: reword this).
+
+## Specification
+
+#### Context Variables
+
+| Name | Type | Meaning |
+| ---- | ------- | --- |
+| `evmmax_state` | `EVMMAXState` | a variable representing ephemeral state which exists for the duration of the current call and in the scope of the current call |
+
+```
+class EVMMAXState():
+	def __init__(self, mod: int, r_squared: int, mod_inv: int, mod_inv_small: int, limb_count: int):
+		self.mod = mod
+		self.r_squared = r_squared
+		self.mod_inv = mod_inv
+		self.mod_inv_small = mod_inv_small
+		self.limb_count
+```
+
+#### Conventions
+
+| Syntax | Meaning |
+| ------ | ------  |
+| `x === y % m` | equality of residue classes: `x % m == y % m` |
+| `pow(x, -1, m)` | modular multiplicative inverse of `x` with respect to modulus `m`: `x * pow(x, -1, m) === -1 % m` |
+
+Introduce a new EOF section kind:
+
+| Name | Value |
+| ---- | ----- |
+| `EVMMAX_MOD_PARAMS` | 3 |
+
+Introduce new opcodes:
+
+| Name | Opcode | Stack In | Stack out | Immediate Size (bytes) |
+| ----- | ----  | -------- | --------- | -------------- |
+| SETMODX | 0x21 | 2 | 0 | 0 |
+| ADDMODX | 0x22 | 0 | 0 | 3 |
+| SUBMODX | 0x23 | 0 | 0 | 3 |
+| MULMONTX | 0x24 | 0 | 0 | 3 |
+| TOMONTX | 0x25 | 2 | 0 | 0 |
+
+`ADDMODX`, `SUBMODX` and `MULMONTX` take inputs encoded in a required 3-byte immediate value appended to the opcode.  Jumpdest analysis is modified to treat the immediate bytes similarly to push-data, disallowing dynamic jumps into them.
+
+### Contract Execution Changes
+
+Upon start of execution (TODO define when exactly this happens wrt various gas checks), if the executing contract code has a `EVMMAX_MOD_PARAMS` section, decode it to an `EVMMAXState` object and associate it with `evmmax_state`:
+
+```
+def decode_mod_params(mod_params_bytes) -> EVMMAXState:
+    # first two bytes are the mod limb count
+    # mod = mod_params_bytes[2:2 + mod_limb_count * 8]
+    offset += mod_limb_count * 8
+    # mod_inv = mod_params_bytes[offset:offset + elem_size]
```

Comment:
This is not needed

Reply:
Can you elaborate? Are you saying that it should be represented more concisely in the spec?

Reply:
mod_inv is not needed for Montgomery conversion or compute, only mod_inv_small (the 64-bit value is).

Though if it is wanted I forgot that you can compute in O(log(log(n))) as it is a power of 2, see https://github.com/mratsim/constantine/blob/master/constantine/math/arithmetic/limbs_exgcd.nim#L44-L77

Explanation p11 "Dumas iterations" based on Newton-Raphson:

- Cetin Kaya Koc (2017), https://eprint.iacr.org/2017/411
- Jean-Guillaume Dumas (2012), https://arxiv.org/pdf/1209.6626v2.pdf
- Colin Plumb (1994), http://groups.google.com/groups?selm=1994Apr6.093116.27805%40mnemosyne.cs.du.edu
Other sources:

- https://crypto.stackexchange.com/questions/47493/how-to-determine-the-multiplicative-inverse-modulo-64-or-other-power-of-two
- https://mumble.net/~campbell/2015/01/21/inverse-mod-power-of-two
- http://marc-b-reynolds.github.io/math/2017/09/18/ModInverse.html